Refactor ST Mbed targets to be CMake buildsystem targets

rwalton-arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Refactor all ST targets to be CMake buildsystem targets. This removes
the need for checking MBED_TARGET_LABELS repeatedly and allows us to be
more flexible in the way we include MBED_TARGET source in the build.

A side effect of this is it will allow us to support custom targets
without breaking the build for 'standard' targets, as we use CMake's
standard mechanism for adding build rules to the build system, rather
than implementing our own layer of logic to exclude files not needed for
the target being built. Using this approach, if an MBED_TARGET is not
linked to using `target_link_libraries` its source files will not be
added to the build. This means custom target source can be added to the
user's application CMakeLists.txt without polluting the build system
when trying to compile for a standard MBED_TARGET.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
You will need to remove any calls to `mbed_set_mbed_target_linker_script`
in your application level `CMakeLists.txt` as this function has been removed.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
This PR depends on PRs to fix the examples' CMakeList files. I'll link the 
PRs here after they've been raised.

### Documentation <!-- Required -->
None
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@Patater @hugueskamba  @0xc0170 @urutva 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

rwalton-arm committed:
cmake: Remove whitespace from mbed-os/CMakeLists.txt

rwalton-arm committed:
cmake: Refactor mbed_set_mbed_target_linker_script

Move the mbed_set_mbed_target_linker_script function into its own file
and make it set the linkerscript on the target passed in, rather than
setting the linkerscript on the top level mbed-core target. We use
CMake's automatic forwarding of usage requirements to forward the linker
script to the top level target.

This commit also adds a condition to check whether the MBED_TARGET is a
CMake buildsystem target. If it is we will just link to it in the
idiomatic CMake way, if it's not then we use the 'old' mechanism of
setting the linkerscript for the mbed-core target. This check will be
useful when we refactor MBED_TARGETS to be CMake buildsystem targets as
it will allow us to do the refactoring incrementally.

Patater commented on tools/cmake/SetLinkerScript.cmake:
```diff
@@ -0,0 +1,48 @@
+                ${LINKER_SCRIPT_PATH}
+            PRE_LINK
+            COMMAND
+                ${CMAKE_C_COMPILER} ${_linker_preprocess_definitions}
```
I like this, using the C compiler in cpp mode, instead of hardcoding the name of the cpp program. Should we use `-x assembler-with-cpp` or `-x c`? Probably doesn't matter, and I don't think there is any `-x linkerscript-with-cpp` option that would be a better description.

multiplemonomials replied:
^ This is the approach I used on mbed-cmake ([here](https://github.com/USCRPL/mbed-cmake/blob/b62358042db62407a686a75ec145599985330808/cmake/MbedOS-GCCArm.cmake#L20)).

jeromecoutant reviewed (approved):
Compilation is OK with 3 random boards.

targets/CMakeLists.txt is not updated as well ?

Patater commented on tools/cmake/SetLinkerScript.cmake:
```diff
@@ -0,0 +1,48 @@
+        target_link_options(${input_target}
+            INTERFACE
+                "--scatter=${raw_linker_script_path}"
+                "--predefine=${_linker_preprocess_definitions}"
```
This `--predefine` option sure saves a lot of work that we have to go through with GCC.

0xc0170 commented on targets/TARGET_STM/TARGET_STM32F0/TARGET_STM32F091xC/CMakeLists.txt:
```diff
@@ -13,15 +11,19 @@ elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
 )
 
-target_include_directories(mbed-core
+target_include_directories(STM32F091xC
```
I assume this follows the naming convection we have in targets.json? do we need to keep this as mixed uppercase with lower case? 
I would use lower case in all targets if possible (the translation should be straightforward from old to new). Or not?

rwalton-arm replied:
This is the convention from the folder structure and targets.json. Generally CMake style is to use camel case for targets, which I see we don't follow in Mbed, so I just went with the folder naming convention. Is the Mbed style to use all lowercase?

0xc0170 commented on tools/cmake/SetLinkerScript.cmake:
```diff
@@ -0,0 +1,48 @@
+#
```
add license info

rwalton-arm replied:
Fixed in [this force push](https://github.com/ARMmbed/mbed-os/compare/99614a04adb48aa444af67bea3a40d6f42a0ac58..7e50d55f8c01fdcd2735eb36d4600cddbbe17d88)

0xc0170 reviewed (commented):
Awesome! We are getting slowly there, CMake targets in `targets` directory ! More to go!

> [x] Major update (Breaking change E.g. Return code change / API behaviour change)

Can we make this nonbreaking ? It would simplify things, I don't think we are ready to release a new major version the next month.

what if we keep the old function there, and print a warning that the function disappears in the next major? We will accumulate more things like this. Or do we need to have the linker in this PR ?

another one: this is porting the targets keeping the name. If we are creating something new, can we fix the naming to follow the components names? 
to illustrate: `NUCLEO_F072RB` would be `nucleo-f072rb` ? Either this or we rename all the components however the application ones use mbed-os/mbed-os-baremetal so I guess rather to follow them than otherway around. To translate this from old names to new should be fairly straightforward (find and replace function in CMake?).

Shall we also create `mcu` targets ? Previous tools had the limitation there to use `TARGET_` as no `MCU_` prefix was implemented. Thus target `MCU_STM32L4R9xI` was using actually TARGET_STM32L4R9xI`.  This is a base target, shouldn't it be named as such? The custom target would inherit from `mcu` targets. The tree would look like `nucleo-f32l4` inherits from `mcu-stm32l4`. I had so many discussions about this problem in the tree, we started adding mcu too late and never had an opportunity to fix it. We might have now.

What about naming scheme for targets, would it make sense to have `mbed-target` prefix? these are specific to mbed, they are targets. What naming should custom targets use? For instance, we used to have `target-frmd-k64f-gcc` in Mbed 3.

Patater commented on targets/TARGET_STM/TARGET_STM32F0/STM32Cube_FW/CMakeLists.txt:
```diff
@@ -1,7 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
-target_sources(mbed-core
+add_library(STM32F0Cube_FW INTERFACE)
```
~Should this be STM32Cube_FW?~ OK, because there are multiple folders named `STM32Cube_FW`, we can't just use the folder name and have to add the family prefix like `F0` in the target name. (Targets names need to be unique.)

rwalton-arm committed:
cmake: Refactor ST targets

Refactor all ST targets to be CMake buildsystem targets. This removes
the need for checking MBED_TARGET_LABELS repeatedly and allows us to be
more flexible in the way we include MBED_TARGET source in the build.

A side effect of this is it will allow us to support custom targets
without breaking the build for 'standard' targets, as we use CMake's
standard mechanism for adding build rules to the build system, rather
than implementing our own layer of logic to exclude files not needed for
the target being built. Using this approach, if an MBED_TARGET is not
linked to using `target_link_libraries` its source files will not be
added to the build. This means custom target source can be added to the
user's application CMakeLists.txt without polluting the build system
when trying to compile for a standard MBED_TARGET.

Patater commented on targets/TARGET_STM/CMakeLists.txt:
```diff
@@ -1,42 +1,29 @@
-elseif("STM32WB" IN_LIST MBED_TARGET_LABELS)
-    add_subdirectory(TARGET_STM32WB)
-endif()
+add_subdirectory(TARGET_STM32F0 EXCLUDE_FROM_ALL)
```
Why `EXCLUDE_FROM_ALL` for `add_subdirectory()` rather than the actual target (`add_library()`)?

rwalton-arm replied:
I guess if a subdirectory happened to add multiple targets they would be all excluded from the ALL target using this approach. However, the main reason is because this is the approach we've taken in other places in Mbed. Is there a good reason to change it?

Patater replied:
Excluding adding the target is the semantically correct thing to do. We want to add all the directories unconditionally, to learn about all their targets. Then, we want to select which target to build/link with. We've so far in Mbed OS decided not to build all targets (CMake default behavior) when the ALL target is built, but instead to build only what isn't explicitly excluded (which is everything except the app). To fix this, we would remove EXCLUDE_FROM_ALL in Mbed OS, and specify which target to build (either from Mbed CLI 2 or directly when using CMake). As a side benefit, having an ALL target helps us to test building all possible board targets easily (just build the ALL target).

Patater replied:
Fixing the ALL target is out-of-scope for this PR, though.

0xc0170 replied:
Please create a story for this work so we can work on this.

rwalton-arm replied:
> Excluding adding the target is the semantically correct thing to do. We want to add all the directories unconditionally, to learn about all their targets. Then, we want to select which target to build/link with. We've so far in Mbed OS decided not to build all targets (CMake default behavior) when the ALL target is built, but instead to build only what isn't explicitly excluded (which is everything except the app). To fix this, we would remove EXCLUDE_FROM_ALL in Mbed OS, and specify which target to build (either from Mbed CLI 2 or directly when using CMake). As a side benefit, having an ALL target helps us to test building all possible board targets easily (just build the ALL target).

Yes, I was aware of the broken ALL target. However, I don't think using `EXCLUDE_FROM_ALL` to exclude the library targets individually is semantically correct either: we're misusing `EXCLUDE_FROM_ALL` currently as a workaround for our misconfigured ALL target and because `mbed-tools` invokes CMake poorly (it always tries to build ALL). So, moving `EXCLUDE_FROM_ALL` from `add_subdirectory` to `add_library` when we're just going to remove it altogether later seems unnecessary to me.

Patater replied:
Agreed

Patater commented on targets/TARGET_STM/TARGET_STM32F3/TARGET_STM32F334x8/CMakeLists.txt:
```diff
@@ -9,14 +9,18 @@ elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
+
+mbed_set_linker_script(STM32F334x8 ${CMAKE_CURRENT_SOURCE_DIR}/${LINKER_FILE})
+
+target_link_libraries(STM32F334x8 INTERFACE STM32F334x8)
```
```diff
- target_link_libraries(STM32F334x8 INTERFACE STM32F334x8)
+ target_link_libraries(STM32F334x8 INTERFACE STM32F3)
```

Patater commented on targets/TARGET_STM/TARGET_STM32G4/TARGET_STM32G484xE/CMakeLists.txt:
```diff
@@ -9,14 +9,16 @@ elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
         .
 )
+
+mbed_set_linker_script(STM32G484xE ${CMAKE_CURRENT_SOURCE_DIR}/${LINKER_FILE})
```
This file is missing a `target_link_libraries()` line.

```diff
+
+ target_link_libraries(STM32G484xE INTERFACE STM32G4)
```

hugueskamba commented on CMakeLists.txt:
```diff
@@ -111,6 +111,32 @@ add_subdirectory(features EXCLUDE_FROM_ALL)
+# preventing CMake from working its usage requirements magic and forcing us to set the linker
+# script globally.
+#
+# TODO: Remove when all MBED_TARGETS have been converted to build system targets.
```
I am guessing the `if` block would still be needed.
```suggestion
# TODO: Remove `else` block when all MBED_TARGETS have been converted to build system targets.
```

rwalton-arm replied:
I have no idea why the `if` block would be needed. All "mbed targets" will be actual buildsystem targets when we remove this. If someone passes in an incorrect target name we should just fail the build. There's no reason to add a check for that as CMake will check it for us.

0xc0170 commented on CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@
 cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
 
 include(${MBED_CONFIG_PATH}/mbed_config.cmake)
+include(tools/cmake/SetLinkerScript.cmake)
```
can we rename to `set_linker_script` ? we do not use camel case naming just CMakeLists does.

0xc0170 commented on CMakeLists.txt:
```diff
@@ -67,7 +67,7 @@ target_compile_definitions(mbed-core
 # Add compile definitions for backward compatibility with the toolchain
 # supported. New source files should instead check for __GNUC__ and __clang__
-# for the GCC_ARM and ARM toolchains respectively. 
+# for the GCC_ARM and ARM toolchains respectively.
```
how this spaces got in... Jamie proposed to have formatter for CMake, we should review that one to avoid these in future.

rwalton-arm replied:
I've used https://github.com/cheshirekow/cmake_format before and it worked well. There are `pre-commit` hooks available, so we could add it to CI quite easily. It's very easy to miss trailing whitespace if your editor isn't set up correctly to flag it.

0xc0170 replied:
I found the PR adding format: https://github.com/ARMmbed/mbed-os/pull/14011 ,if you have a moment, please review

ladislas replied:
besides that, I can make a PR to add .editorconfig -- it's almost a standard and it's super useful for this kind of case.

0xc0170 replied:
> besides that, I can make a PR to add .editorconfig -- it's almost a standard and it's super useful for this kind of case.

That might be useful.

0xc0170 reviewed (dismissed):
The outstanding questions are:

- can we be consistent in CMake to use lower cases with `-` delimeter (translation in CMake via a simple string and replace), no change in tagets.json or tools, just we keep CMake target naming same

Future wish: I wish to have `mcu`, `vendor` or similar in the target names to describe better what is it - and also better hierarchy of targets. This might wait as it needs to refactor targets.json and much more thinking how to structure them (it's not trivial task). we live what we have for now, targets.json

Patater reviewed (dismissed):
LGTM
